common/src/main/kotlin/streams/utils/JSONUtils.kt

```diff
@@ -121,6 +178,100 @@ class DriverRelationshipSerializer : JsonSerializer<org.neo4j.driver.types.Relat
+    }
+
+    override fun convertPayload(payloadMap: JsonNode): RelationshipPayload {
+        return JSONUtils.convertValue<RelationshipPayload>(payloadMap)
+    }
+
+    override fun fillPayload(payload: RelationshipPayload,
+                             beforeProps: Map<String, Any>?,
+                             afterProps: Map<String, Any>?): RelationshipPayload {
+        return payload.copy(
+            before = payload.before?.copy(properties = beforeProps),
+            after = payload.after?.copy(properties = afterProps)
+        )
+    }
+
+    override fun deserialize(parser: JsonParser, context: DeserializationContext): StreamsTransactionRelationshipEvent {
+        val deserialized = super.deserialize(parser, context)
+        if (deserialized.payload.type == EntityType.node) {
+            throw IllegalArgumentException("Relationship event expected, but node type found")
+        }
+        return deserialized
+    }
+
+}
+
+class StreamsTransactionNodeEventDeserializer : StreamsTransactionEventDeserializer<StreamsTransactionNodeEvent, NodePayload>() {
+    override fun createEvent(meta: Meta, payload: NodePayload, schema: Schema): StreamsTransactionNodeEvent {
+        return StreamsTransactionNodeEvent(meta, payload, schema)
+    }
+
+    override fun convertPayload(payloadMap: JsonNode): NodePayload {
+        return JSONUtils.convertValue<NodePayload>(payloadMap)
+    }
+
+    override fun fillPayload(payload: NodePayload,
+                             beforeProps: Map<String, Any>?,
+                             afterProps: Map<String, Any>?): NodePayload {
+        return payload.copy(
+            before = payload.before?.copy(properties = beforeProps),
+            after = payload.after?.copy(properties = afterProps)
+        )
+    }
+
+    override fun deserialize(parser: JsonParser, context: DeserializationContext): StreamsTransactionNodeEvent {
+        val deserialized = super.deserialize(parser, context)
+        if (deserialized.payload.type == EntityType.relationship) {
+            throw IllegalArgumentException("Node event expected, but relationship type found")
+        }
+        return deserialized
+    }
+
+}
+
+abstract class StreamsTransactionEventDeserializer<EVENT, PAYLOAD: Payload> : JsonDeserializer<EVENT>() {
+
+    abstract fun createEvent(meta: Meta, payload: PAYLOAD, schema: Schema): EVENT
+    abstract fun convertPayload(payloadMap: JsonNode): PAYLOAD
+    abstract fun fillPayload(payload: PAYLOAD,
+                             beforeProps: Map<String, Any>?,
+                             afterProps: Map<String, Any>?): PAYLOAD
+
+    @Throws(IOException::class, JsonProcessingException::class)
+    override fun deserialize(parser: JsonParser, context: DeserializationContext): EVENT {
+        val root: JsonNode = parser.codec.readTree(parser)
+        val meta = JSONUtils.convertValue<Meta>(root["meta"])
+        val schema = JSONUtils.convertValue<Schema>(root["schema"])
```

Comment:
am I correct to think that if the user did not include schema information, this still won't work? If so, what would the user do to create point values from messages generated by non-neo4j applications? Regardless, I think we need to extend the documentation accordingly.

Reply:
I would not advocate to the users to use this message created by themselves otherwise we need to document all the possible permutations of schema and so on. So for me, this is about the CDC message which contains the info. If it doesn't it does mean that the message is not well-formed.
Maybe is worth adding docs or at least a link to the CDC doc page (https://neo4j.com/docs/kafka-streams/producer/#_transaction_event_handler) in the Kafka Connect PR where there is no direct reference to the message structure.

Reply:
Yes, looks like we are also missing the type for `geo` properties on the sample messages on the page you linked. Would you please update that page and also a separate page (probably best to avoid links) that includes similar sample messages for the kafka connect documentation?